### ISSUE TITLE
Optimistic UI duplicated update

### ISSUE THREAD

**Issue by zyh41** [deadcoder0904/react-native-darkmode-list#1]
iOS 11.3
RN 0.55
Appsync with Apolllo 2.1

Hey Akshay, I run into the same situation about doing optimistic ui in ios. My issue is that 
1. my seperate home screen which query all items will not get the negative id of optimistic ui from create item
2. in `update` property of my createItem mutation, my optimistic ui negative number return twice too, but eventually the 3rd time is the right one

Any progress on your repo and any clue to fix? Thank you!

**Comment by deadcoder0904** [comment]
No actually Apollo team is bad at answering questions. Its been more than a month or so but still no response or answer.

Keep an eye on https://stackoverflow.com/questions/50603994/apollo-optimistic-ui-does-not-work-in-mutation-component

Also, checkout https://spectrum.chat/thread/4e16ba7d-97a2-40cb-ba2f-bbd22f2228dc

Unfortunately, this same code I've done in React works perfectly fine https://github.com/deadcoder0904/react-darkmodelist 

So I guess this must be an issue with Apollo in combo with React Native but not sure :)

**Comment by zyh41** [comment]
Hey, I think I figured out my problem. My code has no much different than yours. I pulled down your repo. I think the problems is that somehow your server response doesn't update your optimistic response. Instead, it treats as two different items in your `update` function (first I saw a number, then the real response concatenate it the list). Therefore, on your `ListItem` side there are two same items with differenrt ids.

(I would not do the type check if number because in that case you are only updating your UI with optimistic response and won't get the server response). DB is updated correctly so that I think the reason is front end side:

Here are my thoughts you could try:
Change the id from number to string, I think maybe your gql defined id as string so that when it comes back, apollo caches stores treats it as two different items?

**Comment by deadcoder0904** [comment]
The problem with Duplicate Update is solved by me by doing that `typeof` check on https://github.com/deadcoder0904/react-native-darkmode-list/blob/master/App.js#L106. The problem with Optimistic UI is not solved because of that.

As you can see my React code which is same works 👉  https://github.com/deadcoder0904/react-darkmodelist without the `typeof` check on https://github.com/deadcoder0904/react-darkmodelist/blob/master/src/index.js#L61 & if I remove the same `typeof` check from React Native, it results in duplicate item.

If I add the `typeof` check it doesn't show Optimistic UI. 

> Change the id from number to string, I think maybe your gql defined id as string so that when it comes back, apollo caches stores treats it as two different items?

I tried this at first. It produced duplicate items that's why I made `negativeRandom` function as told by @mxstbr on Spectrum (https://spectrum.chat/thread/4e16ba7d-97a2-40cb-ba2f-bbd22f2228dc)

So if you figure out that your React Native app works without producing Duplicate Item & produces Optimistic UI, please point me to your Repo or send a PR if its working fine. 

For now, I am thinking this might be an issue with Apollo in combo with React Native or it might be an issue with Render Prop thing (I have to check it by making it HOC but haven't gotten around yet).

TL;DR

- React repo works fine without `typeof` check & it does not give Duplicate items & it also provides Optimistic UI
- React Native repo produces duplicate items without `typeof` check
- React Native repo does not give Optimistic UI. It only prints after receiving response from server

**Comment by deadcoder0904** [comment]
Apparently it was a bug in Apollo or React Apollo in combo with React Native so I just updated my dependencies & it worked. Also, removed `typeof` check plus `negativeRandom` function as now Optimistic UI works out of the box & now my code is similar to the React one. Check out my latest commit for more info :)
